core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/batchmanager/BatchBuffer.java

```diff
@@ -0,0 +1,146 @@
+     * response pair is received.
+     */
+    private final AtomicInteger nextId;
+
+    /**
+     * Keeps track of the ID of the next entry to be added in a batch request. This ID does not necessarily correlate to a
+     * request that already exists in the idToBatchContext map since it refers to the next entry (ex. if the last entry added
+     * to idToBatchContext had an id of 22, nextBatchEntry will have a value of 23).
+     */
+    private final AtomicInteger nextBatchEntry;
+
+    /**
+     * The scheduled flush tasks associated with this batchBuffer.
+     */
+    private ScheduledFuture<?> scheduledFlush;
+
+    public BatchBuffer(ScheduledFuture<?> scheduledFlush) {
+        this.idToBatchContext = new ConcurrentHashMap<>();
+        this.nextId = new AtomicInteger(0);
+        this.nextBatchEntry = new AtomicInteger(0);
+        this.scheduledFlush = scheduledFlush;
+    }
+
+    public Map<String, BatchingExecutionContext<RequestT, ResponseT>> canManualFlush(int maxBatchItems) {
+        synchronized (flushLock) {
+            if (idToBatchContext.size() >= maxBatchItems) {
+                return extractFlushedEntries(maxBatchItems);
+            }
+            return null;
+        }
+    }
+
+    public Map<String, BatchingExecutionContext<RequestT, ResponseT>> canScheduledFlush(int maxBatchItems) {
+        synchronized (flushLock) {
+            if (idToBatchContext.size() > 0) {
+                return extractFlushedEntries(maxBatchItems);
+            }
+            return null;
+        }
+    }
+
+    private Map<String, BatchingExecutionContext<RequestT, ResponseT>> extractFlushedEntries(int maxBatchItems) {
+        LinkedHashMap<String, BatchingExecutionContext<RequestT, ResponseT>> requestEntries = new LinkedHashMap<>();
+        String nextEntry;
+        while (requestEntries.size() < maxBatchItems && (nextEntry = nextBatchEntry()) != null) {
+            requestEntries.put(nextEntry, idToBatchContext.get(nextEntry));
+            idToBatchContext.remove(nextEntry);
+        }
+        return requestEntries;
+    }
+
+    public RequestT getRequest(String key) {
+        return idToBatchContext.get(key).request();
+    }
+
+    public CompletableFuture<ResponseT> getResponse(String key) {
+        return idToBatchContext.get(key).response();
+    }
+
+    // TODO: Needs to be in a lock to maintain insertion order. Not sure if there is any other way to accomplish this. I tried to
+    //  do this in a do while loop but it still ended up resulting in an incorrect insertion order.
+    public BatchingExecutionContext<RequestT, ResponseT> put(RequestT request, CompletableFuture<ResponseT> response) {
+        synchronized (this) {
+            String id = BatchUtils.getAndIncrementId(nextId);
+            return idToBatchContext.put(id, new BatchingExecutionContext<>(request, response));
+        }
+    }
+
+    private String nextBatchEntry() {
```

Comment:
Do we still need nextBatchEntry?

Reply:
Replaced with a nextBatchEntry() and hasNextBatchEntry() method. Need to figure out better names to differentiate between nextBatchEntry and nextId.